File: android/src/main/java/ch/nevis/mobile/authentication/sdk/react/methods/userInteraction/IsPolicyCompliantMethodHandler.kt
/*
 * Copyright © 2023 Nevis Security AG. All rights reserved.
 */

package ch.nevis.mobile.authentication.sdk.react.methods.userInteraction

import ch.nevis.mobile.authentication.sdk.react.MethodChannelHandler
import ch.nevis.mobile.authentication.sdk.react.cache.OperationCache
import ch.nevis.mobile.authentication.sdk.react.cache.operation.UserInteractionOperation
import ch.nevis.mobile.authentication.sdk.react.cache.state.UserInteractionState
import ch.nevis.mobile.authentication.sdk.react.methods.MethodHandler
import ch.nevis.mobile.authentication.sdk.react.model.ReactMethodName
import ch.nevis.mobile.authentication.sdk.react.model.message.incoming.ChannelInMessage
import ch.nevis.mobile.authentication.sdk.react.model.message.incoming.IsPolicyCompliantMessage
import ch.nevis.mobile.authentication.sdk.react.model.message.outgoing.IsPolicyCompliantOutMessage
import ch.nevis.mobile.sdk.api.MobileAuthenticationClient

class IsPolicyCompliantMethodHandler : MethodHandler() {
    override fun execute(
        client: MobileAuthenticationClient,
        channelInMessage: ChannelInMessage,
    ) {
        val message = validate<IsPolicyCompliantMessage>(channelInMessage)
        val currentState =
            OperationCache.getOperation<UserInteractionOperation>(message.operationId).state as? UserInteractionState
                ?: throw IllegalStateException("isPolicyCompliant() can only be called during selectAuthenticator or selectAccount!")
        val resultMessage =	IsPolicyCompliantOutMessage(
            message.operationId,
            currentState.isPolicyCompliant(
                message.aaid,
                message.username
            )
        )

        MethodChannelHandler.resolve(ReactMethodName.IS_POLICY_COMPLIANT, resultMessage)
    }
}

File: ios/Methods/UserInteraction/IsPolicyCompliantMethodHandler.swift
//
// Copyright © 2023 Nevis Security AG. All rights reserved.
//

import NevisMobileAuthentication

struct IsPolicyCompliantMethodHandler: MethodHandler {
	func execute(using _: MobileAuthenticationClient, with message: ChannelInMessage) throws {
		let message: IsPolicyCompliantMessage = try validate(message: message)
		let operation: UserInteractionOperation = try OperationCache.shared.read(by: message.operationId)
		guard let state = operation.state as? UserInteractionState else {
			throw PluginError.illegalState("Select Account or Authenticator state is expected!")
		}

		let isPolicyCompliant = state.isPolicyCompliant(aaid: message.aaid,
		                                                for: message.username)
		let resultMessage = IsPolicyCompliantOutMessage(operationId: message.operationId, isPolicyCompliant: isPolicyCompliant)
		MethodChannelHandler.shared.resolve(method: .isPolicyCompliant, message: resultMessage)
	}
}
